src/diagnostics.rs

```diff
@@ -11,12 +11,24 @@ use lsp_types::*;
 use std::collections::HashMap;
 use std::fmt::Write as _;
 
-pub fn publish_diagnostics(params: Params, ctx: &mut Context) {
+pub fn publish_diagnostics(server_name: &ServerName, params: Params, ctx: &mut Context) {
     let params: PublishDiagnosticsParams = params.parse().expect("Failed to parse params");
     let path = params.uri.to_file_path().unwrap();
     let buffile = path.to_str().unwrap();
-    ctx.diagnostics
-        .insert(buffile.to_string(), params.diagnostics);
+    let mut diagnostics: Vec<_> = ctx
+        .diagnostics
+        .remove(buffile)
+        .unwrap_or_default()
+        .into_iter()
+        .filter(|(name, _)| name != server_name)
+        .collect();
+    let params: Vec<_> = params
+        .diagnostics
+        .into_iter()
+        .map(|d| (server_name.clone(), d))
+        .collect();
+    diagnostics.extend(params);
+    ctx.diagnostics.insert(buffile.to_string(), diagnostics);
```

Comment:
I guess this would be more elegant if we used `ctx.diagnostics.entry().or_insert()...` instead of remove() followed by insert(). Didn't try it though